### PR TITLE
Functions metrics prometheus

### PR DESCRIPTION
**Pull request by jerrypeng**
add functions metrics look like the following:

pulsar_function__total_processed__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_processed__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_processed__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_processed__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_successfully_processed__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_successfully_processed__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_successfully_processed__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_successfully_processed__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_system_exceptions__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_system_exceptions__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_system_exceptions__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_system_exceptions__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_user_exceptions__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_user_exceptions__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_user_exceptions__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_user_exceptions__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_serialization_exceptions__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_serialization_exceptions__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_serialization_exceptions__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__total_serialization_exceptions__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__avg_latency_ms__count{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__avg_latency_ms__max{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__avg_latency_ms__min{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950
pulsar_function__avg_latency_ms__sum{cluster="standalone", namespace="test/test-namespace", name="test1", instanceId="0"} 0.0 1520920466950

### CODE REVIEW

**Comment by jerrypeng** [comment]
@srkukarni @sijie @merlimat please review

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                                String tenant = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getTenant();
+                                String namespace = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getNamespace();
+                                String name = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getName();
+                                int instanceId = functionRuntimeInfo.getFunctionInstance().getInstanceId();
+                                String qualifiedNamespace = String.format("%s/%s", tenant, namespace);
+
+                                metric(out, cluster, qualifiedNamespace, name, String.format("pulsar_function%scount", metricName),
```
Comment (srkukarni) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: maybe _ before count

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/JavaInstanceMain.java [hunk 2]
```diff
@@ -221,5 +222,22 @@ public void getFunctionStatus(Empty request, StreamObserver<InstanceCommunicatio
                 throw new RuntimeException(e);
             }
         }
+
```
Comment (srkukarni) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/JavaInstanceMain.java]: arent we removing metrics collection from spawner?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+package org.apache.pulsar.broker.stats.prometheus;
```
Comment (srkukarni) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: Does this file belong here?

Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: Well the other metrics generators are here.  The generator doesn't collect metrics its just puts them into prometheus format

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java (hunk 1 above)
Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: The metric name already container "__" before and after e.g. __total_processed__

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/JavaInstanceMain.java (hunk 2 above)
Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/JavaInstanceMain.java]: Sure but I would say lets do it in multiple PRs

**Comment by srkukarni** [comment]
retest this please

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java (hunk 3 above)
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: I think this file is missing license header?

Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: good catch I will add

Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/FunctionsStatsGenerator.java]: I have add the missing license header

**Comment by sijie** [comment]
retest this please

**Comment by jerrypeng** [comment]
retest

**Comment by jerrypeng** [comment]
@srkukarni @sijie @merlimat can you review the PR again.  I refactored as per our discussion

**Comment by jerrypeng** [comment]
@srkukarni I have added the instance liveness check

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeSpawner.java [hunk 4]
```diff
@@ -40,23 +41,39 @@
...
         this.numRestarts = 0;
+        this.instanceLivenessCheckFreqMs = instanceLivenessCheckFreqMs;
     }
 
     public void start() throws Exception {
         log.info("RuntimeSpawner starting function {} - {}", this.instanceConfig.getFunctionConfig().getName(),
                 this.instanceConfig.getInstanceId());
         runtime = runtimeFactory.createContainer(this.instanceConfig, codeFile);
         runtime.start();
+        // monitor function runtime to make sure it is running.  If not, restart the function runtime
```
Comment (srkukarni) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeSpawner.java]: shouldnt we check for null?

**Comment by srkukarni** [comment]
looks good

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionsStatsGenerator.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                                String metricName = metricsEntry.getKey();
+                                InstanceCommunication.MetricsData.DataDigest dataDigest = metricsEntry.getValue();
+
+                                String tenant = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getTenant();
+                                String namespace = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getNamespace();
+                                String name = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getName();
+                                int instanceId = functionRuntimeInfo.getFunctionInstance().getInstanceId();
```
Comment (merlimat) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionsStatsGenerator.java]: We should have some test to verify the metrics output. There is already a parser for the Prometheus format in the tests code. It parses the stats into a map. Take a look at `PrometheusMetricsTest`

Comment (jerrypeng) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionsStatsGenerator.java]: sounds good. I will add a test

**Comment by jerrypeng** [comment]
@merlimat thanks for the review. I have added a unit test

**Comment by merlimat** [comment]
retest this please

**Comment by jerrypeng** [comment]
retest this please

**Pull request merged**